caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -309,71 +309,111 @@ public boolean isFile( ConcreteResource resource )
                         {
-                            String errorMsg = "";
-                            if ( e instanceof IOException )
-                            {
-                                errorMsg = String.format(
-                                        "[galley] got i/o error when doing the NFS->Local copy for resource %s",
-                                        resource.toString() );
-                            }
-                            else
-                            {
-                                errorMsg = String.format(
-                                        "[galley] Cache TransactionManager got error, locking key is %s, resource is %s",
-                                        pathKey, resource.toString() );
-                            }
+                            final String errorMsg = String.format(
+                                    "[galley] Cache TransactionManager got error, locking key is %s, resource is %s",
+                                    pathKey, resource.toString() );
                             logger.error( errorMsg, e );
                             throw new IllegalStateException( errorMsg, e );
                         }
-                        finally
+                    }
+                    finally
+                    {
+                        try
+                        {
+                            cacheTxMgr.rollback();
+                        }
+                        catch ( SystemException e )
                         {
-                            try
-                            {
-                                cacheTxMgr.rollback();
-                            }
-                            catch ( SystemException e )
-                            {
-                                final String errorMsg = String.format(
-                                        "[galley] Cache TransactionManager rollback got error, locking key is %s",
-                                        pathKey );
-                                logger.error( errorMsg, e );
-                            }
+                            final String errorMsg = String.format(
+                                    "[galley] Cache TransactionManager rollback got error, locking key is %s", pathKey );
+                            logger.error( errorMsg, e );
                         }
+                        IOUtils.closeQuietly( nfsIn );
+                        IOUtils.closeQuietly( localOut );
+                        canStreamOpen.set( true );
+                        copyLock.notifyAll();
                     }
-                } );
+                }
+            }
+        };
+        // This lock is used to control the concurrent operations on the resource, like concurrent delete and read/write.
+        // Use "this" as lock is heavy, should think about use the transfer for the resource as the lock for each thread
+        synchronized ( getTransfer( resource ) )
```

Comment:
Something tells me combining this synchronized block with the nested one on copyLock may be leading to the deadlock you mentioned in your comments at the top of the PR.